# Ridge and Lasso Regression - Lab

## Introduction

In this lab, you'll practice your knowledge on Ridge and Lasso regression!

## Objectives

You will be able to:

- Use Lasso and ridge regression in Python
- Compare Lasso and Ridge with standard regression

## Recall our cost functions

From previously, you know that when solving for a linear regression, you can express the cost function as

This is the expression for simple linear regression (for 1 predictor $x$). If you have multiple predictors, you would have something that looks like:

$$ \text{cost_function}= \sum_{i=1}^n(y_i - \hat{y})^2 = \sum_{i=1}^n(y_i - \sum_{j=1}^k(m_jx_{ij} + b))^2$$

where $k$ is the number of predictors.

In ridge regression, the linear regression cost function is changed by adding a penalty term to square of the magnitude of the coefficients.

$$ \text{cost_function_ridge}= \sum_{i=1}^n(y_i - \hat{y})^2 = \sum_{i=1}^n(y_i - \sum_{j=1}^k(m_jx_{ij} + b))^2 + \lambda \sum_{j=1}^p m_j^2$$

$$ \text{cost_function_lasso}= \sum_{i=1}^n(y_i - \hat{y})^2 = \sum_{i=1}^n(y_i - \sum_{j=1}^k(m_jx_{ij} + b))^2 + \lambda \sum_{j=1}^p \mid m_j \mid$$

Note that, for our gradients, when having multiple predictors $x_j$ with $j \in 1,\ldots, k$

$$ \frac{dJ}{dm_j}J(m_j,b) = -2\sum_{i = 1}^n x_{j,i}(y_i - (\sum_{j=1}^km_j{x_{ij}} + b)) = -2\sum_{i = 1}^n x_{ij}*\epsilon_i$$
$$ \frac{dJ}{db}J(m_j,b) = -2\sum_{i = 1}^n(y_i - (\sum_{j=1}^km_j{x_{ij}} + b)) = -2\sum_{i = 1}^n \epsilon_i $$
    

## An example using our `auto-mpg` data

Let's transform our continuous predictors in `auto-mpg` and see how they perform as predictors in a Ridge versus Lasso regression.

In [5]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import Lasso, Ridge, LinearRegression
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

data = pd.read_csv("auto-mpg.csv") 

In [6]:
data

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0         130    3504          12.0   
1    15.0          8         350.0         165    3693          11.5   
2    18.0          8         318.0         150    3436          11.0   
3    16.0          8         304.0         150    3433          12.0   
4    17.0          8         302.0         140    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
387  27.0          4         140.0          86    2790          15.6   
388  44.0          4          97.0          52    2130          24.6   
389  32.0          4         135.0          84    2295          11.6   
390  28.0          4         120.0          79    2625          18.6   
391  31.0          4         119.0          82    2720          19.4   

     model year  origin                   car name  
0            70       1  chevrolet chevelle malibu  
1            70       1          buick skylark 320  
2            70       1         plymouth satellite  
3            70       1              amc rebel sst  
4            70       1                ford torino  
..          ...     ...                        ...  
387          82       1            ford mustang gl  
388          82       2                  vw pickup  
389          82       1              dodge rampage  
390          82       1                ford ranger  
391          82       1                 chevy s-10  

[392 rows x 9 columns]

In [7]:
data['horsepower'].astype(str).astype(int)
y = data[["mpg"]]
X = data.drop(["mpg", "car name", "origin"], axis=1)

scale = MinMaxScaler()
transformed = scale.fit_transform(X)
X = pd.DataFrame(transformed, columns = X.columns)

data= pd.concat([y,X], axis=1, ignore_index= True)

data.head()

0    1         2         3         4         5    6
0  18.0  1.0  0.617571  0.456522  0.536150  0.238095  0.0
1  15.0  1.0  0.728682  0.646739  0.589736  0.208333  0.0
2  18.0  1.0  0.645995  0.565217  0.516870  0.178571  0.0
3  16.0  1.0  0.609819  0.565217  0.516019  0.238095  0.0
4  17.0  1.0  0.604651  0.510870  0.520556  0.148810  0.0

Below, we created train-test-splits, and created Ridge, Lasso and Linear regression models

In [8]:
# Perform test train split
X_train , X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=12)

In [9]:
data_train = pd.concat([y_train,X_train], axis=1, ignore_index= True)

In [10]:
data_train

0    1         2         3         4         5         6
44   18.0  0.6  0.490956  0.347826  0.382478  0.327381  0.083333
334  27.2  0.2  0.173127  0.206522  0.248653  0.458333  0.916667
224  20.5  0.6  0.421189  0.320652  0.513751  0.529762  0.583333
355  30.7  0.6  0.198966  0.163043  0.438616  0.690476  0.916667
11   14.0  1.0  0.702842  0.619565  0.565920  0.000000  0.000000
..    ...  ...       ...       ...       ...       ...       ...
130  25.0  0.2  0.186047  0.157609  0.263397  0.535714  0.333333
241  21.5  0.0  0.031008  0.347826  0.313864  0.327381  0.583333
253  25.1  0.2  0.186047  0.228261  0.313864  0.440476  0.666667
155  15.0  1.0  0.728682  0.538043  0.801531  0.357143  0.416667
331  23.7  0.0  0.005168  0.293478  0.228806  0.267857  0.833333

[274 rows x 7 columns]

In [11]:
data_train = data_train.reset_index(drop=True).T

In [12]:
data_train

0          1          2          3          4          5    \
0  18.000000  27.200000  20.500000  30.700000  14.000000  19.000000   
1   0.600000   0.200000   0.600000   0.600000   1.000000   0.600000   
2   0.490956   0.173127   0.421189   0.198966   0.702842   0.405685   
3   0.347826   0.206522   0.320652   0.163043   0.619565   0.266304   
4   0.382478   0.248653   0.513751   0.438616   0.565920   0.468103   
5   0.327381   0.458333   0.529762   0.690476   0.000000   0.476190   
6   0.083333   0.916667   0.583333   0.916667   0.000000   0.416667   

         6          7          8          9    ...        264        265  \
0  19.000000  24.000000  15.000000  14.000000  ...  26.000000  28.000000   
1   0.600000   0.200000   0.600000   1.000000  ...   0.200000   0.200000   
2   0.470284   0.186047   0.470284   0.728682  ...   0.074935   0.124031   
3   0.228261   0.250000   0.293478   0.646739  ...   0.173913   0.239130   
4   0.478877   0.354976   0.488517   0.736036  ...   0.194783   0.144599   
5   0.446429   0.500000   0.535714   0.238095  ...   0.386905   0.357143   
6   0.083333   1.000000   0.333333   0.083333  ...   0.333333   0.083333   

         266        267        268        269        270        271  \
0  21.500000  35.000000  18.100000  25.000000  21.500000  25.100000   
1   0.600000   0.200000   0.600000   0.200000   0.000000   0.200000   
2   0.421189   0.139535   0.490956   0.186047   0.031008   0.186047   
3   0.375000   0.228261   0.402174   0.157609   0.347826   0.228261   
4   0.462716   0.251489   0.509498   0.263397   0.313864   0.313864   
5   0.440476   0.422619   0.422619   0.535714   0.327381   0.440476   
6   0.750000   0.833333   0.666667   0.333333   0.583333   0.666667   

         272        273  
0  15.000000  23.700000  
1   1.000000   0.000000  
2   0.728682   0.005168  
3   0.538043   0.293478  
4   0.801531   0.228806  
5   0.357143   0.267857  
6   0.416667   0.833333  

[7 rows x 274 columns]

In [13]:
def step_gradient(b_current, m_current ,points):
    b_gradient = 0
    m_gradient = np.zeros(len(m_current))
    learning_rate = .01
    N = float(len(points))
    for i in range(0, len(points)):
        y = points[i][0]
        x = points[i][1:(len(m_current)+1)] 
        b_gradient += -(1/N)  * (y -  (sum(m_current * x) + b_current))
        m_gradient += -(1/N) * x * (y -  (sum(m_current * x) + b_current))
    new_b = b_current - (learning_rate * b_gradient)
    new_m = m_current - (learning_rate * m_gradient)
    return (new_b, new_m)

In [14]:
len(data_train)

7

In [15]:
data_train[0][0] #y

18.0

In [16]:
data_train[0][1:7] #x

1    0.600000
2    0.490956
3    0.347826
4    0.382478
5    0.327381
6    0.083333
Name: 0, dtype: float64

In [21]:
b = 0
m = [0,0,0,0,0,0]
updated_b, updated_m = step_gradient(b, m, data_train) 
# {'b': 0.0085, 'm': 0.6249999999999999}

In [22]:
updated_b

0.212

In [23]:
updated_m

1    0.119657
2    0.078246
3    0.059325
4    0.090801
5    0.097067
6    0.108619
Name: 0, dtype: float64

In [24]:
# set our initial step with m and b values, and the corresponding error.
b = 0
m = [0,0,0,0,0,0]
iterations = []
for i in range(5000):
    iteration = step_gradient(b, m, data_train)
    b= iteration[0]
    m = []
    for j in range(len(iteration[1])):
        m.append(iteration[1][j+1])
    iterations.append(iteration)

In [25]:
iterations[4999]

(12.746290723761629, 1     1.571408
 2    -0.922501
 3    -0.640380
 4     0.519102
 5     7.499974
 6    10.899802
 Name: 0, dtype: float64)

In [26]:
# Build a Ridge, Lasso and regular linear regression model. 
# Note how in scikit learn, the regularization parameter is denoted by alpha (and not lambda)
ridge = Ridge(alpha=0.5)
ridge.fit(X_train, y_train)

lasso = Lasso(alpha=0.5)
lasso.fit(X_train, y_train)

lin = LinearRegression()
lin.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

Next, let's create predictions for train and test sets.

In [28]:
# Create preditions for training and test sets
y_h_ridge_train = ridge.predict(X_train)
y_h_ridge_test = ridge.predict(X_test)

y_h_lasso_train = np.reshape(lasso.predict(X_train), (274,1))
y_h_lasso_test = np.reshape(lasso.predict(X_test), (118,1))

y_h_lin_train = lin.predict(X_train)
y_h_lin_test = lin.predict(X_test)

Look at the RSS for train and test for each of the three models.

In [29]:
print('Train Error Ridge Model', np.sum((y_train - y_h_ridge_train)**2))
print('Test Error Ridge Model', np.sum((y_test - y_h_ridge_test)**2))
print('\n')

print('Train Error Lasso Model', np.sum((y_train - y_h_lasso_train)**2))
print('Test Error Lasso Model', np.sum((y_test - y_h_lasso_test)**2))
print('\n')

print('Train Error Unpenalized Linear Model', np.sum((y_train - lin.predict(X_train))**2))
print('Test Error Unpenalized Linear Model', np.sum((y_test - lin.predict(X_test))**2))

Train Error Ridge Model mpg    2688.222824
dtype: float64
Test Error Ridge Model mpg    2074.197775
dtype: float64


Train Error Lasso Model mpg    4644.536425
dtype: float64
Test Error Lasso Model mpg    3696.183375
dtype: float64


Train Error Unpenalized Linear Model mpg    2658.043444
dtype: float64
Test Error Unpenalized Linear Model mpg    1976.266987
dtype: float64


We note that Ridge is clearly better than Lasso here, but that the unpenalized model performs best here. Let's see how including Ridge and Lasso changed our parameter estimates.

In [30]:
print('Ridge parameter coefficients:', ridge.coef_)
print('Lasso parameter coefficients:', lasso.coef_)
print('Linear model parameter coefficients:', lin.coef_)

Ridge parameter coefficients: [[ -2.11792413  -3.0112953   -1.90579654 -15.60758962  -1.61071692
    8.12940111]]
Lasso parameter coefficients: [-10.31005725  -0.          -0.          -2.27967948   0.
   3.88327477]
Linear model parameter coefficients: [[ -1.33790698  -1.05300843  -0.08661412 -20.08143923  -0.39639115
    8.56051229]]


You can clearly see how Lasso shrinks certain parameters to 0! The Ridge regression mostly affected the fourth parameter (estimated to be -20.08 for the linear regression model).

## Additional reading

Full code examples for Ridge and Lasso regression, advantages and disadvantages, and how to code ridge and Lasso in Python can be found [here](https://www.analyticsvidhya.com/blog/2016/01/complete-tutorial-ridge-lasso-regression-python/).

Make sure to have a look at the Scikit-Learn documentation for [Ridge](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html) and [Lasso](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html).


## Summary

Great! You now know how to perform Lasso and Ridge regression. Let's move on to the lab to explore Lasso and Ridge further!